### PR TITLE
Change reachability module to use StableMIR

### PR DESCRIPTION
**Pull request by celinval**
Rewrite the reachability module to use Stable APIs wherever possible. Note that in StableMIR the instance body is already monomorphized and constants are already evaluated, which simplifies the code for most of it, except to handle stubbing issue #2589.

For the stubbing issue, we still use a mix of stable and internal APIs to detect an invalid monomorphization.

### Call-outs

1. ~~This change requires the newest version of the Rust toolchain. Thus, it is upgrading the version as well.~~
2. ~~Creating this as a draft until https://github.com/rust-lang/rust/pull/118274 is merged.~~
3. I ran s2n-quic manually to assess performance and no significant difference.
  - s2n-quic-core:  36.86 vs 37.19
  - s2n-quic-platform: 16.17 vs 16.19

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: tests/cargo-kani/stubbing-double-extern-path/harness/expected [hunk 1]
```diff
@@ -1 +1 @@
-error[E0391]: cycle detected when optimizing MIR for `crate_a::assert_true`
```
Comment (celinval) [inline tests/cargo-kani/stubbing-double-extern-path/harness/expected:1]: FYI, this test checks that there is a cycle so the error printed here depends on which method we visit first. But the cycle is correctly detected no matter what.

File: kani-compiler/src/kani_middle/reachability.rs [hunk 2]
```diff
@@ -13,146 +13,135 @@
...
+        // Filter regular items.
+        crate_items
+            .iter()
+            .filter_map(|item| {
+                // Only collect monomorphic items.
+                Instance::try_from(*item).ok().and_then(|instance| {
+                    let def_id = rustc_internal::internal(item);
+                    predicate(tcx, def_id)
+                        .then_some(InternalMonoItem::Fn(rustc_internal::internal(&instance)))
+                })
```
Comment (adpaco-aws) [inline kani-compiler/src/kani_middle/reachability.rs:92]: I'm not able to establish how this corresponds to what we had. What does `internal` do?

File: kani-compiler/src/kani_middle/reachability.rs [hunk 3]
```diff
@@ -13,146 +13,135 @@
...
+    collected: FxHashSet<MonoItem>,
     /// Items enqueued for visiting.
-    queue: Vec<MonoItem<'tcx>>,
+    queue: Vec<MonoItem>,
     #[cfg(debug_assertions)]
-    call_graph: debug::CallGraph<'tcx>,
+    call_graph: debug::CallGraph,
 }
 
+// TODO: Need to figure out how to handle stubbing errors.
```
Comment (adpaco-aws) [inline kani-compiler/src/kani_middle/reachability.rs:144]: What's this referring to? Please open an issue (unless we have it) and refer to it here.

File: kani-compiler/src/kani_middle/stubbing/mod.rs [hunk 4]
```diff
@@ -28,3 +35,83 @@ pub fn harness_stub_map(
...
     stub_pairs
 }
+
+/// Validate that an instance body can be instantiated.
+///
+/// Stubbing may cause an instance to not be correctly instantiated since we delay checking its
+/// generic bounds.
+///
+/// In stable mir, trying to retrieve an Instance::body() will ICE if we cannot evaluate a
+/// constant as expected. For now, use internal APIs to anticipate this issue.
```
Comment (adpaco-aws) [inline kani-compiler/src/kani_middle/stubbing/mod.rs:45]:
```suggestion
/// In stable MIR, trying to retrieve an `Instance::body()` will ICE if we cannot evaluate a
/// constant as expected. For now, use internal APIs to anticipate this issue.
```

File: kani-compiler/src/kani_middle/reachability.rs [hunk 5]
```diff
@@ -194,146 +184,117 @@ impl<'tcx> MonoItemsCollector<'tcx> {
...
-    fn collect_vtable_methods(&mut self, concrete_ty: Ty<'tcx>, trait_ty: Ty<'tcx>) {
+    fn collect_vtable_methods(&mut self, concrete_ty: Ty, trait_ty: Ty) {
         trace!(?concrete_ty, ?trait_ty, "collect_vtable_methods");
-        assert!(!concrete_ty.is_trait(), "Expected a concrete type, but found: {concrete_ty:?}");
-        assert!(trait_ty.is_trait(), "Expected a trait: {trait_ty:?}");
-        if let TyKind::Dynamic(trait_list, ..) = trait_ty.kind() {
+        let concrete_kind = concrete_ty.kind();
+        let trait_kind = trait_ty.kind();
+
+        assert!(!concrete_kind.is_trait(), "Expected a concrete type, but found: {concrete_ty:?}");
```
Comment (adpaco-aws) [inline kani-compiler/src/kani_middle/reachability.rs:244]:
```suggestion
        assert!(!concrete_kind.is_trait(), "expected a concrete type, but found `{concrete_ty:?}`");
```

File: kani-compiler/src/kani_middle/reachability.rs [hunk 6]
```diff
@@ -194,146 +184,117 @@ impl<'tcx> MonoItemsCollector<'tcx> {
...
+        assert!(trait_kind.is_trait(), "Expected a trait: {trait_ty:?}");
```
Comment (adpaco-aws) [inline kani-compiler/src/kani_middle/reachability.rs:245]:
```suggestion
        assert!(trait_kind.is_trait(), "expected a trait `{trait_ty:?}`");
```

File: kani-compiler/src/kani_middle/reachability.rs [hunk 7]
```diff
@@ -518,22 +422,22 @@ impl<'a, 'tcx> MirVisitor<'tcx> for MonoItemsFnCollector<'a, 'tcx> {
...
                                 // even the return type, for instance for a
                                 // trait like `FromIterator`.
-                                let generic_ty = outer_args[0].ty(self.body, tcx).peel_refs();
-                                let receiver_ty = tcx.instantiate_and_normalize_erasing_regions(
-                                    substs,
-                                    ParamEnv::reveal_all(),
-                                    EarlyBinder::bind(generic_ty),
-                                );
+                                let receiver_ty = args.0[0].expect_ty();
+                                trace!(?receiver_ty, ?callee, "**** receiver:");
```
Comment (adpaco-aws) [inline kani-compiler/src/kani_middle/reachability.rs:426]: Looks much simpler than what we used to do. Do we want to keep the `trace!` call?

**Review by adpaco-aws — COMMENTED** [review]
It's looking really good. I wonder if we could add links to the stable MIR project or its APIs so it's clear what's going on here. But I'm not sure how much you'll use these APIs in other parts of the compiler, maybe it's better to add that elsewhere.

File: kani-compiler/src/kani_middle/reachability.rs (hunk 7 above)
Comment (celinval) [reply kani-compiler/src/kani_middle/reachability.rs:426]: No. Also a debug artifact that I need to remove. :)

**Comment by zhassan-aws** [comment]
> I wonder if we could add links to the stable MIR project or its APIs so it's clear what's going on here.

I second this request. It's a bit difficult for me to understand the changes given my lack of knowledge of how the stable/internal MIR APIs compare.

**Comment by celinval** [comment]
> > I wonder if we could add links to the stable MIR project or its APIs so it's clear what's going on here.
> 
> I second this request. It's a bit difficult for me to understand the changes given my lack of knowledge of how the stable/internal MIR APIs compare.

Sure. Right now `stable_mir` and `rustc_smir` are new crates in the rust compiler itself. So there's nothing new, and you can apply the same logic as you currently use to check APIs from `rustc_span` and `rustc_middle`.

I will add more information to the code and to [`main.rs`](https://github.com/model-checking/kani/blob/main/kani-compiler/src/main.rs), since this code will start showing up in different parts of our compiler.

File: kani-compiler/src/kani_middle/reachability.rs (hunk 2 above)
Comment (celinval) [reply kani-compiler/src/kani_middle/reachability.rs:92]: This code collects all items in the current crate (`all_local_items`), and tries to instantiate them (via `Instance::try_from()`). Only the non-generic code will be successfully converted.

For those that were non-generic, we convert the stable Instance to the internal version, so we can invoke the closure that was passed as an argument which still expects the internal instance.

**Comment by celinval** [comment]
@zhassan-aws and @adpaco-aws, I added some documentation to our wiki instead. I hope that's what you were looking for. Cheers!

File: docs/src/stable_mir.md [hunk 8]
```diff
@@ -0,0 +1,94 @@
+# Transition to StableMIR
```
Comment (zhassan-aws) [inline docs/src/stable_mir.md:1]: Very detailed and informative! Thanks for writing this.

**Review by zhassan-aws — APPROVED** [review]

**Pull request merged**